core/jvm/src/main/scala/fs2/interop/flow/StreamSubscriber.scala

```diff
@@ -38,230 +37,288 @@ import java.util.concurrent.atomic.AtomicReference
-
-    sealed trait Input
-    case class OnSubscribe(s: Subscription) extends Input
-    case class OnNext(a: A) extends Input
-    case class OnError(e: Throwable) extends Input
-    case object OnComplete extends Input
-    case object OnFinalize extends Input
-    case class OnDequeue(response: Out => Unit) extends Input
-
-    sealed trait State
-    case object Uninitialized extends State
-    case class Idle(sub: Subscription, buffer: Chunk[A]) extends State
-    case class RequestBeforeSubscription(req: Out => Unit) extends State
-    case class WaitingOnUpstream(
-        sub: Subscription,
-        buffer: Chunk[A],
-        elementRequest: Out => Unit
-    ) extends State
-    case object UpstreamCompletion extends State
-    case object DownstreamCancellation extends State
-    case class UpstreamError(err: Throwable) extends State
-
-    def reportFailure(e: Throwable): Unit =
-      Thread.getDefaultUncaughtExceptionHandler match {
-        case null => e.printStackTrace()
-        case h    => h.uncaughtException(Thread.currentThread(), e)
+        case state =>
+          Failed(new InvalidStateException(operation = s"Received record [${a}]", state)) -> noop
       }
 
-    def step(in: Input): State => (State, () => Unit) =
-      in match {
-        case OnSubscribe(s) => {
-          case RequestBeforeSubscription(req) =>
-            WaitingOnUpstream(s, Chunk.empty, req) -> (() => s.request(chunkSize.toLong))
+      case Error(ex) => {
+        case Uninitialized(Some(cb)) =>
+          Terminal -> run {
+            cb.apply(Left(ex))
+          }
 
-          case Uninitialized =>
-            Idle(s, Chunk.empty) -> (() => ())
+        case WaitingOnUpstream(_, _, cb, _) =>
+          Terminal -> run {
+            cb.apply(Left(ex))
+          }
+
+        case _ =>
+          Failed(new UpstreamErrorException(ex)) -> noop
+      }
+
+      case Complete(canceled) => {
+        case Uninitialized(Some(cb)) =>
+          Terminal -> run {
+            cb.apply(Right(None))
+          }
 
-          case o =>
-            val err = new Error(s"Received subscription in invalid state [${o}]")
-            o -> { () =>
+        case Idle(s) =>
+          Terminal -> run {
+            if (canceled) {
               s.cancel()
-              reportFailure(err)
             }
-        }
+          }
 
-        case OnNext(a) => {
-          case WaitingOnUpstream(s, buffer, r) =>
-            val newBuffer = buffer ++ Chunk.singleton(a)
-            if (newBuffer.size == chunkSize)
-              Idle(s, Chunk.empty) -> (() => r(newBuffer.some.asRight))
-            else
-              WaitingOnUpstream(s, newBuffer, r) -> (() => ())
+        case WaitingOnUpstream(idx, buffer, cb, s) =>
+          Terminal -> run {
+            if (idx == 0) {
+              cb.apply(Right(None))
+            } else {
+              cb.apply(Right(Some(Chunk.array(buffer, offset = 0, length = idx))))
```

Comment:
How does the `Stream` terminate in this case? Since it needs a `None` to terminate.

Reply:
After the next pull, it receives the `None` :)
I would love to send that `None` right away, but we need another callback for that.